#92 Update Nuke documentations Install&Plugin (#93)
Inbibo/shift_docs: docs/integration_resources/software/nuke.md

## Shift installation in Nuke

To install Shift and open it through a menu entry within Nuke, it is required to set up Shift in the environment or in the user *init.py* Python files from Nuke. Shift can be added to the user *init.py* file from Nuke's preference folder (*<home directory>/.nuke*) or to a custom *init.py* file inside a custom folder structure. If the directory where these files are stored is a path sourced by Nuke, they will be automatically executed on startup.

### Init File/Environment requirements

You can add these requirements to your environment before opening Nuke:

<pre><code style="white-space: pre; margin: 20px 0; padding: 10px; box-sizing: border-box;">
PATH=&#37;PATH&#37;&semi;"&ltpath_to_the_shift_installation_folder&gt"

> PATH=&#37;PATH&#37;&semi;"&ltpath_to_the_shift_installation_folder&gt/shift/thirdparty/python/Lib/site-packages"

Reviewer: Aren't these two PYTHONPATH instead of PATH???

Author: right, fixed

> <pre><code style="white-space: pre; margin: 20px 0; padding: 10px; box-sizing: border-box;">
> PATH=&#37;PATH&#37;&semi;"&ltpath_to_the_shift_installation_folder&gt"
> PATH=&#37;PATH&#37;&semi;"&ltpath_to_the_shift_installation_folder&gt/shift/thirdparty/python/Lib/site-packages"
> NUKE_PATH="&ltpath_to_the_shift_installation_folder&gt/shift/plugins/nuke/startup"&semi;&#37;NUKE_PATH&#37;
> 

Reviewer 2: To remove unnecessary spaces
```suggestion
<pre><code style="white-space: pre; margin: 20px 0; padding: 10px; box-sizing: border-box;">PATH=&#37;PATH&#37;&semi;"&ltpath_to_the_shift_installation_folder&gt"
PATH=&#37;PATH&#37;&semi;"&ltpath_to_the_shift_installation_folder&gt/shift/thirdparty/python/Lib/site-packages"
NUKE_PATH="&ltpath_to_the_shift_installation_folder&gt/shift/plugins/nuke/startup"&semi;&#37;NUKE_PATH&#37;
```

</code></pre>

> Or you can add them to the *init.py* file to make Shift to work in Nuke batch and interactive modes. To achieve this, it is only required to add the Shift installation path to the **PATH** environment variable.

Reviewer: This is not exactly correct as appending to sys.path does not change the environment variable PATH but only the python resources.
```suggestion
Or you can add them to the *init.py* file to source Shift in Nuke batch and interactive modes. To achieve this, it is only required to add the Shift installation path to the Python `sys.path`.

```

> <pre><code style="white-space: pre; margin: 20px 0; padding: 10px; box-sizing: border-box;">
> import nuke

Reviewer 2: To remove upper space
```suggestion
<pre><code style="white-space: pre; margin: 20px 0; padding: 10px; box-sizing: border-box;">import nuke

```

try:
    import shift
except:
    import sys
    sys.path.append("&ltpath_to_the_shift_installation_folder&gt")
    sys.path.append("&ltpath_to_the_shift_installation_folder&gt/shift/thirdparty/python/Lib/site-packages")

nuke.pluginAddPath("&ltpath_to_the_shift_installation_folder&gt/shift/plugins/nuke/startup")
</code></pre>

>[!NOTE]
> If the paths are added at the system level or before opening Nuke, then this configuration step will not be required.

## Shift in Nuke

> When the requirements are added you will Shift a Shift menu entry in the toolbar and top menu inside Nuke.

Reviewer 2: ```suggestion
When requirements are added, a Shift menu entry will appear in both the toolbar and top menu in Nuke.

```

This entry will contain the access to open the Shift UI and to create a ShiftWorkflow Node.

<figure>
      <img src="images/nuke_shift_toolbar.png" alt="Shift Toolbar">
      <figcaption><b>Figure 1</b>: Shift tools in Nuke nodes menu.</figcaption>
</figure>

<figure>
      <img src="images/nuke_shift_menu.png" alt="Shift menu">
      <figcaption><b>Figure 2</b>: Shift UI on Nuke's top menu bar.</figcaption>
</figure>

> you will be able to create ShiftWorkflow nodes usign the tab key in the node graph too.

Reviewer 2: ```suggestion
You can create ShiftWorkflow nodes by using the tab key in the node graph as well.

```

<figure>
      <img src="images/nuke_shift_search_node.png" alt="Creating ShiftWorkflow node.">
      <figcaption><b>Figure 3</b>: Create ShiftWorkflow nodes in node graph window.</figcaption>
</figure>

## ShiftWorkflow Node

> The Shift plugin for Nuke includes a node to execute Shift workflow's in the Nuke's node graph.
> 
> In the Shiftworkflow node you can select you Shift workflow to load it in the node and automatically generate the required input and output knobs and inputs. (If auto plugs option is checked.)

Reviewer 2: ```suggestion
The Shift plugin for Nuke includes a node to execute a Shift workflow within Nuke's node graph.

In the Shiftworkflow node, a Shift workflow can be selected to load it into the node. This will automatically generate the required input and output knobs and connections, if the auto plugs option is enabled

```

Reviewer: please add a `.` at the end

<figure>
      <img src="images/nuke_shift_node_propierties.png" alt="ShiftWorkflow propierties.">
      <figcaption><b>Figure 3</b>: Main Propierties tab of ShiftWorkflow node.</figcaption>
</figure>

You can use relative paths if you use the **SHIFT_PATH_WORKFLOWS** env variable.

> ### Plugs <--> Knobs/Inputs

Reviewer 2: I would change this title.
Somthing like Shift Plugs to Nuke Knobs/Inputs

Reviewer: agree this does nt show correctly in the website
![image](https://github.com/user-attachments/assets/b7c3f48b-0a7b-4b37-b548-81d0c3ac5d86)

Author: done

> This is the correlation of Shift Plugs to Nuke knobs or inputs.

Reviewer: ```suggestion
The following table shows the correspondance between Shift Plugs and Nuke knobs or inputs.

```

> This is the correlation of Shift Plugs to Nuke knobs or inputs.

Reviewer: Please add an introduction describing the fact that the Nuke plugin will read the workflow plugs from Input and Output operators. Stress on the fact that functionalities of the plugin rely on the naming of certain plugs chosen at creation time. Renaming might cause issues with the detection.

Moreover the `*` formatting is not clear. Please state in a sentence the formatting meaning for plug detection

Author: Note added.

Reviewer: Added a comment here https://github.com/Inbibo/shift_docs/pull/93/files#r1882324057

Please review it

| SPlug Type   | SPlug Code     | Nuke Type                                        | Knob Type                                                    | Note                                                                                                |
|:-------------|:---------------|:-------------------------------------------------|:-------------------------------------------------------------|:----------------------------------------------------------------------------------------------------|
| *Bool*       | -              | Knob                                             | Boolean_Knob                                                 |                                                                                                     |
| *Color*      | -              | Knob                                             | AColor_Knob                                                  |                                                                                                     |

> | *Dict*       | -              | Knob                                             | EvalString_Knob                                              | Require custom implementation                                                                       |

Reviewer 2: ```suggestion
| *Dict*       | -              | Knob                                             | EvalString_Knob                                              | Requires custom implementation                                                                       |

```

| *FileIn*     | -              | Knob                                             | File_Knob                                                    |                                                                                                     |
| *FileOut*    | -              | Knob                                             | File_Knob                                                    |                                                                                                     |
| *Float*      | -              | Knob                                             | Double_Knob                                                  |                                                                                                     |
| *Instance*   | image*         | Input/Output<br/>Knob<br/>Knob                   | - <br/>Channel_Knob <br/>File_Knob                           | Provides the input image like a single image. (Current Frame)                                       |

> | *Instance*   | images*        | Output<br/>Knob<br/>Knob                         | - <br/>File_Knob                                             | Provides the input image like a single image. (Current Frame)                                       |

Reviewer: Why images is only output??

Let's open a conversation on this durnig a meeting.

Reviewer: After discussing this is set as output only because the "images" are lists of images which are not time correlated and for this reason we do not intend the user to use them as inputs.

This type of plug is useful for ease of displaying multiple examples produced by a workflow.

> | *Instance*   | images*        | Output<br/>Knob<br/>Knob                         | - <br/>File_Knob                                             | Provides the input image like a single image. (Current Frame)                                       |

Reviewer: Please update the Note on this type of plug

```suggestion
| *Instance*   | images*        | Output<br/>Knob<br/>Knob                         | - <br/>File_Knob                                             | Provides the input image like a single image. (Current Frame)                                       |

```

Reviewer: Please review the suggestion here https://github.com/Inbibo/shift_docs/pull/93/files#r1882338295

Author: commited

> | *Instance*   | imageSequence* | Input/Output<br/>Knob<br/>Knob<br/>Knob<br/>Knob | - <br/>Channel_Knob <br/>File_Knob<br/>Int_Knob<br/>Int_Knob | Provides the input image like a image sequence rendering the full frame range provided in the knob. |

Reviewer 2: ```suggestion
| *Instance*   | imageSequence* | Input/Output<br/>Knob<br/>Knob<br/>Knob<br/>Knob | - <br/>Channel_Knob <br/>File_Knob<br/>Int_Knob<br/>Int_Knob | Provides the input image like an image sequence rendering the full frame range provided in the knob. |

```

> | *Instance*   | imageSequence* | Input/Output<br/>Knob<br/>Knob<br/>Knob<br/>Knob | - <br/>Channel_Knob <br/>File_Knob<br/>Int_Knob<br/>Int_Knob | Provides the input image like a image sequence rendering the full frame range provided in the knob. |

Reviewer: This is an implementation stress point. PIL instances are picklable, therefore we do not need these to be kInstance plugs. They can be kObject types as well.

Reviewer: After discussing in a meeting this is the implementation allowed for the moment. We will revisit this when we decide what to do with the image plugs. resolving

| *Instance*   | -              | Knob                                             | EvalString_Knob                                              | Require custom implementation                                                                       |
| *Int*        | -              | Knob                                             | Int_Knob                                                     |                                                                                                     |

> | *List*       | points*        | Input                                            | -                                                            | Requires to connect a Tracker node to get the points from.                                          |

Reviewer: Add the knob for y coord switching and describe in the Note for this plug

> | *Instance*   | -              | Knob                                             | EvalString_Knob                                              | Require custom implementation                                                                       |
> | *Int*        | -              | Knob                                             | Int_Knob                                                     |                                                                                                     |
> | *List*       | points*        | Input                                            | -                                                            | Requires to connect a Tracker node to get the points from.                                          |
> | *List*       | -              | Knob                                             | EvalString_Knob                                              | Require custom implementation                                                                       |
> | *Object*     | -              | Knob                                             | EvalString_Knob                                              | Require custom implementation                                                                       |

Reviewer 2: ```suggestion
| *Instance*   | -              | Knob                                             | EvalString_Knob                                              | Requires custom implementation                                                                       |
| *Int*        | -              | Knob                                             | Int_Knob                                                     |                                                                                                     |
| *List*       | points*        | Input                                            | -                                                            | Requires to connect a Tracker node to get the points from.                                          |
| *List*       | -              | Knob                                             | EvalString_Knob                                              | Requires custom implementation                                                                       |
| *Object*     | -              | Knob                                             | EvalString_Knob                                              | Requires custom implementation                                                                       |

```

| *String*     | -              | Knob                                             | EvalString_Knob                                              |                                                                                                     |

> ### Node config

Reviewer: ```suggestion
### Node Config

```

The node config have 3 options:
- autoClean: Removes automatically all the temporal renders done for the input when the executiong ends.

> - backupOutputImage: Copy the output image in 

Reviewer: ```suggestion
- backupOutputImage: Make a backup copy the output image.

```

Author: done

> The node config have 3 options:
> - autoClean: Removes automatically all the temporal renders done for the input when the executiong ends.
> - backupOutputImage: Copy the output image in 
> - imageType: Can be PIL or np.array. This knob specifies the image instance type to use to provide and read from the input and output from the workflow. If the workflow expects or return a difference type it may not work.

Reviewer 2: ```suggestion
The node configuration has three options:
- autoClean: Automatically removes all temporary renders created for the input once the execution ends.
- backupOutputImage:  Copies the output image.
- imageType: Specifies the format used to write and read the inputs and outputs for the workflow.  The instance types can be either PIL or np.array.  If the workflow expects or returns a different type, it may not work correctly.
```

> 

Reviewer: Please add a disclaimer html box stressing on the fact that: If PIL is chosen an sRGB image (in PNG format) will be provided, while for np.array case the image will be in Linear space (in EXR format).

Author: added new sections to go deeper on this. Pls review them and let me know if they are enough or something is missing.

      <figcaption><b>Figure 1</b>: Shift tools in Nuke nodes menu.</figcaption>
</figure>

<figure>
      <img src="images/nuke_shift_menu.png" alt="Shift menu">
      <figcaption><b>Figure 2</b>: Shift UI on Nuke's top menu bar.</figcaption>
</figure>

You can create ShiftWorkflow nodes by using the tab key in the node graph as well.

<figure>
      <img src="images/nuke_shift_search_node.png" alt="Creating ShiftWorkflow node.">
      <figcaption><b>Figure 3</b>: Create ShiftWorkflow nodes in node graph window.</figcaption>
</figure>

## ShiftWorkflow Node

The Shift plugin for Nuke includes a node to execute a Shift workflow within Nuke's node graph.

In the Shiftworkflow node, a Shift workflow can be selected to load it into the node. This will automatically generate the required input and output knobs and connections, if the auto plugs option is enabled.

<figure>
      <img src="images/nuke_shift_node_propierties.png" alt="ShiftWorkflow propierties.">
      <figcaption><b>Figure 3</b>: Main Propierties tab of ShiftWorkflow node.</figcaption>
</figure>

You can use relative paths if you use the **SHIFT_PATH_WORKFLOWS** env variable.

> ### Shift Plugs to Nuke Knobs/Inputs
> 
> The following table shows the correspondance between Shift Plugs and Nuke knobs or inputs.
> 

Reviewer: ```suggestion
### Shift Plugs to Nuke Knobs/Inputs

When initialized with a Shift workflow, the node will parse through the Input and Output nodes plugs and automatically populate a set of Nuke knobs in the node UI.

The following table shows the correspondance between Shift Plugs and Nuke knobs or inputs.

```

| *FileIn*     | -              | Knob                                             | File_Knob                                                    |                                                                                                                                                                                                                                                           |
| *FileOut*    | -              | Knob                                             | File_Knob                                                    |                                                                                                                                                                                                                                                           |
| *Float*      | -              | Knob                                             | Double_Knob                                                  |                                                                                                                                                                                                                                                           |

> | *Instance*   | image*         | Input/Output<br/>Knob<br/>Knob                   | - <br/>Channel_Knob <br/>File_Knob                           | Provides the input image like a single image. (Current Frame)                                                                                                                                                                                             |

Reviewer: ```suggestion
| *Instance*   | image*         | Input/Output<br/>Knob<br/>Knob                   | - <br/>Channel_Knob <br/>File_Knob                           | Renders/Loads the input/output image like a single image. (Current Frame)                                                                                                                                                                                             |

```

> | *Instance*   | images*        | Output<br/>Knob<br/>Knob                         | - <br/>File_Knob<br/>Float_Knob(Switch selector)             | For output plugs returns a list of single images that might not be related between them. The user will be able to select which one to display with a Switch.                                                                                              |

Reviewer: ```suggestion
| *Instance*   | images*        | Output<br/>Knob<br/>Knob                         | - <br/>File_Knob<br/>Float_Knob(Switch selector)             | Used for outputting a list of images that are not temporally correlated. The user will be able to select which one to display with a Switch.                                                                                              |

```

> | *Instance*   | imageSequence* | Input/Output<br/>Knob<br/>Knob<br/>Knob<br/>Knob | - <br/>Channel_Knob <br/>File_Knob<br/>Int_Knob<br/>Int_Knob | Provides the input image like an image sequence rendering the full frame range provided in the knob.                                                                                                                                                      |

Reviewer: ```suggestion
| *Instance*   | imageSequence* | Input/Output<br/>Knob<br/>Knob<br/>Knob<br/>Knob | - <br/>Channel_Knob <br/>File_Knob<br/>Int_Knob<br/>Int_Knob | Renders/Loads the input/output images like an image sequence rendering the full frame range provided in the knob.                                                                                                                                                      |

```

| *Object*     | -              | Knob                                             | EvalString_Knob                                              | Requires custom implementation                                                                                                                                                                                                                            |
| *String*     | -              | Knob                                             | EvalString_Knob                                              |                                                                                                                                                                                                                                                           |

> >[!NOTE]
> > * The Plugs to Knobs relations with a SPlug Code required means that the name of the plug need to start by that name, but can be longer. For example you can have more than 1 image with Shift Plug: "image1", "image2",...

Reviewer: Asterisk displays as dotted list in the website, please fix that.

Here is a suggestion for the text, please commit this suggestion prior to the tests for the asterisk display.
```suggestion
>[!NOTE]
> * The conversion for certain Shift Plugs to Nuke Knobs will change depending on the Plug naming protocol. Specifically the shiftWorkflow node plugin for Nuke will identify as "special", plugs that start with the word defined in the `SPlug Code` section of the table above, e.g. "image1", "image2",... or "imageSequenceOriginal", "imageSequenceMasked",...

```

*Image Plugs*

The Image plugs represented like inputs or outputs have a file knob where the user can define the path where the image have to be saved.

> For input images, they will be saved temporally. For output images they will be saved, but override if the workflow is executed again.

Author: The suggestion have duplicated sentences. Check this suggestion and feel free to commit it if it's ok

```suggestion
Input images will be saved as temporary files for them to be loaded in Python. Output images will be saved to the designated path, however successive executions will overwrite them on disk if the path is not changed.

```

> For input images, they will be saved temporally. For output images they will be saved, but override if the workflow is executed again.

Reviewer: Correct, you can commit yours! Thank you!

### Node Config

The node configuration has three options:
- autoClean: Automatically removes all temporary renders created for the input once the execution ends.
- backupOutputImage: Makes a backup copy the output image.
- imageType: Specifies the format used to write and read the inputs and outputs for the workflow.  The instance types can be either PIL or np.array.  If the workflow expects or returns a different type, it may not work correctly.

<figure>
      <img src="images/nuke_shift_node_propierties_mode.png" alt="Image Mode.">
      <figcaption><b>Figure 3</b>: Node config tab.</figcaption>
</figure>

### Image Colorspace

> The ShiftWorkflow node manage the input and output images internally. For input images, a write node is connected to the image input, for output images a Read node is created to load the image back in Nuke.
> By default the correlation of colorspaces and image types is the following:
> 
> | Image Type    | Default Colorspace | Data Type | File Type |
> |:--------------|:-------------------|:----------|:----------|
> | PIL.Image     | sRGB               | 8 bit     | png       |
> | Numpy.Array   | Linear (Raw)       | 32 float  | exr       |
> 
> The colorspace of the input or the output image could be changed in the internal write nodes and Read nodes of the ShiftWorkflow node if desired.
> You can also make color conversions inside your Shift workflow to convert from one colorspace to others.

Reviewer: ```suggestion
The ShiftWorkflow node manages the input and output images internally. For input images, a Write node is connected to the image Input, for output images a Read node is created to load the image back in Nuke.
By default the correlation of colorspaces and image types is the following:

| Image Type    | Default Colorspace | Data Type | File Type |
|:--------------|:-------------------|:----------|:----------|
| PIL.Image     | sRGB               | 8 bit     | png       |
| Numpy.Array   | Linear (Raw)       | 32 float  | exr       |

The colorspace of the input or the output image can be changed in the internal Write nodes and Read nodes of the ShiftWorkflow node if desired.
You can also perform conversions inside your Shift workflow to pass from one colorspace to other.

```

| *Object*     | -              | Knob                                             | EvalString_Knob                                              | Requires custom implementation                                                                                                                                                                                                                            |
| *String*     | -              | Knob                                             | EvalString_Knob                                              |                                                                                                                                                                                                                                                           |

>[!NOTE]

> > * The conversion for certain Shift Plugs to Nuke Knobs will change depending on the Plug naming protocol. Specifically the shiftWorkflow node plugin for Nuke will identify as "special", plugs that start with the word defined in the `SPlug Code` section of the table above, e.g. "image1", "image2",... or "imageSequenceOriginal", "imageSequenceMasked",...

Author: ```suggestion
> &#42; The conversion for certain Shift Plugs to Nuke Knobs will change depending on the Plug naming protocol. Specifically the shiftWorkflow node plugin for Nuke will identify as "special", plugs that start with the word defined in the `SPlug Code` section of the table above, e.g. "image1", "image2",... or "imageSequenceOriginal", "imageSequenceMasked",...

```

Reviewer: Did this work? I don't see it committed to the code

Author: I changed it locally. At the end I pushed \*. I think it´s working

### Node Config

The node configuration has three options:
- autoClean: Automatically removes all temporary renders created for the input once the execution ends.
- backupOutputImage: Makes a backup copy the output image.
- imageType: Specifies the format used to write and read the inputs and outputs for the workflow.  The instance types can be either PIL or np.array.  If the workflow expects or returns a different type, it may not work correctly.

<figure>
      <img src="images/nuke_shift_node_propierties_mode.png" alt="Image Mode.">
      <figcaption><b>Figure 3</b>: Node config tab.</figcaption>
</figure>

### Image Colorspace

The ShiftWorkflow node manages the input and output images internally. For input images, a Write node is connected to the image Input, for output images a Read node is created to load the image back in Nuke.
By default the correlation of colorspaces and image types is the following:

| Image Type    | Default Colorspace | Data Type | File Type |
|:--------------|:-------------------|:----------|:----------|
| PIL.Image     | sRGB               | 8 bit     | png       |

> | Numpy.Array   | Linear (Raw)       | 32 float  | exr       |

Author: commited locally to avoid break the table, but applied like suggested.

## Shift installation in Nuke

To install Shift and open it through a menu entry within Nuke, it is required to set up Shift in the environment or in the user *init.py* Python files from Nuke. Shift can be added to the user *init.py* file from Nuke's preference folder (*<home directory>/.nuke*) or to a custom *init.py* file inside a custom folder structure. If the directory where these files are stored is a path sourced by Nuke, they will be automatically executed on startup.

> ### Init File/Environment requirements

Reviewer 2: ```suggestion
### Init File and Environment requirements

```
